### PR TITLE
feat: mirror create-side validation guards onto update path (closes #1292)

### LINKED ISSUE #1292: [FEATURE] Mirror create-side validation guards to update path in ha_config_set_helper (parity with #1150)
### What would you like to see?

Add the same input-shape validation guards to the `update` branch of `ha_config_set_helper` that currently fire on `create`:

1. **`input_select`** — `initial`-must-be-in-`options` guard (currently at `src/ha_mcp/tools/tools_config_helpers.py:L2405-2422`, missing in the update branch at L3046-3056).
2. **`input_datetime`** — `has_date`/`has_time`-both-False rejection (currently at L2481-2488, missing in the update branch at L3132-3142).

### Why do you need this?

These create-side guards exist because Home Assistant's own error responses on these inputs are generic enough that callers couldn't act on them — the `input_select` guard was specifically motivated by #1150 ("Bug-4a") and produces a typed `ErrorCode.VALIDATION_INVALID_PARAMETER` with `suggestions` instead of letting the caller hit HA's underlying error.

The update branch of the same tool merges values without re-running the guard, so:

- A caller updating `options` and `initial` together (or only `options`, leaving `initial` outside the new option list) reaches the same broken-entity state with `success: true` that #1150 fixed for the create path.
- A caller explicitly passing `has_date=False, has_time=False` on update merges both False and writes the payload to HA, which then surfaces its generic error — exactly the UX the create-side guard exists to avoid.

This is asymmetric application of the project's "tool-side validation when it genuinely adds value" exception under the `.gemini/styleguide.md` → "Validation delegation" principle. The same precedent that justified the create-side guard supports parity on the update path.

### Additional context

The unified `ha_config_set_helper` tool (post-#1012) routes both create and update through the same outer entry point but currently diverges in the per-helper-type branches:

| Helper type | Branch | Line | Guard present? |
|---|---|---|---|
| `input_select` | create | L2405-2422 | ✓ `initial in options` enforced |
| `input_select` | update | L3046-3056 | ✗ values merged, guard absent |
| `input_datetime` | create | L2481-2488 | ✓ `has_date or has_time` enforced |
| `input_datetime` | update | L3132-3142 | ✗ values merged, guard absent |

Suggested fix: extract the create-branch validation blocks into local helper functions and call them from both branches, or duplicate the guards in the update branch — whichever fits the existing code organisation in `tools_config_helpers.py`. Tests for the update path should cover both rejection cases (mirroring whatever exists for the create path under #1150).

### PR DESCRIPTION
**Pull request by Patch76**
## What does this PR do?

Closes the parity gap that issue #1292 surfaces: two input-side validation guards that fire on the **create** branch of `ha_config_set_helper` are absent from the **update** branch.

| Helper type | Constraint | Create-side (existed since #1150) | Update-side (before this PR) |
|---|---|---|---|
| `input_select` | `initial` ∈ `options` | ✓ enforced (`src/ha_mcp/tools/tools_config_helpers.py:L2401-2422` on master) | ✗ values merged, guard absent |
| `input_datetime` | at least one of `has_date` / `has_time` is `True` | ✓ enforced (`L2481-2488` on master) | ✗ values merged, guard absent |

A caller hitting either today reaches the same broken-entity state with `success: true` that #1150 closed for the create path, or surfaces the generic HA error message the create-side guard exists to pre-empt.

**Approach** — extract-and-share, matching the sibling pattern at `tools_config_helpers.py:L869-998` (`_validate_numeric_range`, `_validate_input_select_options`, `_validate_schedule_days`):

- `_validate_initial_in_options(options, initial, helper_type="input_select")` — accepts the resolved values from either branch; treats `initial=None` as the unset case so a merge-fall-through that leaves `initial` unset doesn't false-positive. `options: Any` with an internal `isinstance(options, list)` early-return mirrors the sibling validator's shape-guard pattern.
- `_validate_datetime_has_date_or_time(has_date, has_time)` — flags only the explicit `(False, False)` resolved-after-merge combo the issue targets; `None` passes through (not constrained).

Both validators are called from the create branch (replacing the inline blocks) and from the update branch after the merge step that resolves the final `(options, initial)` and `(has_date, has_time)` pairs.

**User-visible message shift on the create path:** the inline create-branch error message previously suggested *"Or omit `initial` so the entity starts unset"*. The shared validator now reads *"Or omit `initial` to use the default or existing value."* — a generalization for the dual-branch use, accurate on both code paths (omit on create → unset/default; omit on update → preserve existing via the merge).

## Type of change
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 📚 Documentation
- [ ] 🔧 Maintenance/refactor
- [ ] 🧪 Tests only
- [ ] 💥 Breaking change

## Testing
- [ ] I have tested these changes with a LLM agent
- [x] All automated tests pass (`uv run pytest`)
- [x] Code follows style guidelines (`uv run ruff check`)

Local runs (Alpine, Python 3.13 musl):
- **Unit tests**: 2459 passed, 1 skipped (numpy by-design), 0 failed
- **New tests** (`tests/src/unit/test_helper_input_validation.py`): 20 added — 4 closing the previously-untested coverage gap for the #1150 create-side guards, 8 covering the new update-side parity (5 input_select, 3 input_datetime), 5 direct shape-guard contract tests (`TestValidateInitialInOptionsShapeGuard` — None/str/dict options pass through silently; `None` initial with list options passes through; `helper_type` param threads to error context), and 3 edge tests (`TestValidateInitialInOptionsEdges` — `initial=""` against non-empty options rejects; `options=[]` with any initial rejects; end-to-end update path with `initial=""` against existing options rejects).
- **Mypy**: clean on `tools_config_helpers.py`
- **Ruff check**: clean on all touched files

## Future improvements

- **Sibling-validator audit on update paths**: the same create-only-guard pattern that motivated this PR may exist elsewhere in `tools_config_helpers.py` (e.g. `counter`, `timer`, `input_text` ranges). The numeric-range guard already runs on update via `_validate_numeric_range`; the remaining per-type fields would benefit from the same audit. Out-of-scope here to keep the PR narrowly aligned with issue #1292.

## Checklist
- [x] I have updated documentation if needed (no public docs reference these validators)

### CODE REVIEW

File: src/ha_mcp/tools/tools_config_helpers.py [hunk 1]
```diff
@@ -867,6 +867,53 @@ def _validate_numeric_range(
...
+    since that's what reaches HA as the broken-entity payload.
+    """
+    if has_date is False and has_time is False:
+        raise_tool_error(
+            create_error_response(
+                ErrorCode.VALIDATION_INVALID_PARAMETER,
+                "At least one of has_date or has_time must be True for input_datetime",
+                context=_simple_helper_error_context("input_datetime"),
+            )
+        )
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_config_helpers.py:930]: ✅ Fixed in 716d5a9 — added `context={has_date, has_time}` plus a two-entry `suggestions` array. Active-instruction wording ("Set has_date=True...") mirrors the sibling `_validate_initial_in_options` in this PR; documentation-link suggestion from the proposal dropped to keep parity with that sibling (and `_validate_input_select_options`), neither of which ships a docs link.

File: src/ha_mcp/tools/tools_config_helpers.py [hunk 2]
```diff
@@ -867,6 +867,61 @@ def _validate_numeric_range(
...
 
+def _validate_initial_in_options(options: list[Any], initial: Any) -> None:
+    """Reject input_select ``initial`` values not in ``options`` (Bug 4a, issue #1150).
+
+    The create-branch guard (#1150) and the update-branch guard (#1292) call
+    this with the resolved values — caller-supplied or merged from the
+    existing config. ``initial=None`` is the unset case and passes through.
+    """
+    if initial is None:
+        return
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_config_helpers.py]: ✅ Fixed in fef3885 — added `isinstance(options, list)` to the early-return guard, mirroring the defensive shape check in `_validate_input_select_options` (L932). Both current callers (create L2460, update L3091) feed a list, but the guard now prevents a future non-list caller from raising a confusing `TypeError` on `initial not in options`.

File: src/ha_mcp/tools/tools_config_helpers.py [hunk 3]
```diff
@@ -867,6 +867,61 @@ def _validate_numeric_range(
...
+                f"initial={initial!r} must be one of options "
+                f"{options!r} for input_select.",
+                context=_simple_helper_error_context(
+                    "input_select",
+                    initial=initial,
+                    options=options,
+                ),
+                suggestions=[
+                    "Pick an `initial` value that's in `options`.",
+                    "Or omit `initial` so the entity starts unset.",
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_config_helpers.py]: ✅ Fixed in fef3885 — reworded to `"Or omit initial to use the default or existing value."`. Suggestions are now accurate for both the create path (omit → unset/default) and the update path (omit → preserve existing via the merge at L3079-3084).

File: src/ha_mcp/tools/tools_config_helpers.py [hunk 4]
```diff
@@ -867,6 +867,65 @@ def _validate_numeric_range(
...
+                f"{options!r} for input_select.",
+                context=_simple_helper_error_context(
+                    "input_select",
+                    initial=initial,
+                    options=options,
+                ),
+                suggestions=[
+                    "Pick an `initial` value that's in `options`.",
+                    "Or omit `initial` to use the default or existing value.",
+                ],
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_config_helpers.py:900]: 📝 Not addressing — the current wording "default or existing value" tells the caller that omitting falls back to existing; users hitting the edge case (new `options` + existing `initial` outside them) can apply suggestion #1 ("Pick an 'initial' value that's in 'options'"). Three-suggestion arrays with a documentation link aren't the sibling pattern here — `_validate_input_select_options` ships one suggestion and no doc-link, and the Round-1 review on `_validate_datetime_has_date_or_time` settled on no doc-link for the same parity reason.

**Comment by Patch76** [comment]
## Implementation Summary

**Choices Made:**
- **Extract-and-share validators, mirror onto update path** — `_validate_initial_in_options(options, initial)` and `_validate_datetime_has_date_or_time(has_date, has_time)` are reusable helpers called from both create (replacing the inline blocks the #1150 work shipped) and update (after the merge step resolves the final caller-or-existing values). Matches the established sibling pattern at `tools_config_helpers.py:L869-998` for `_validate_numeric_range`, `_validate_input_select_options`, and `_validate_schedule_days`. Closes the parity gap that issue #1292 surfaces: callers used to reach a broken-entity state with `success: true` on update for both `input_select.initial ∉ options` and `input_datetime.(has_date, has_time) == (False, False)`.
- **Datetime validator error response now structured per styleguide** — adopted Gemini's first-wave finding on `_validate_datetime_has_date_or_time`. The error now ships `context={has_date, has_time}` so the failing payload is visible in the structured error, plus a two-entry `suggestions` array in the active-instruction style of `_validate_initial_in_options`. Documentation-link suggestion from the bot proposal dropped to keep parity with the sibling validators (neither `_validate_initial_in_options` nor `_validate_input_select_options` ships one).
- **Defensive isinstance guard on `_validate_initial_in_options`** — adopted Gemini's second-wave L878 finding. Mirrors the existing `isinstance(options, list)` shape guard in `_validate_input_select_options:L932`. Both current callers feed a list (create at L2460, update at L3091, the latter via the `existing.get("options", [])` default), so the guard is a no-op today; it prevents a future non-list caller from raising a confusing `TypeError` on `initial not in options`.
- **Neutral suggestion wording on `_validate_initial_in_options`** — adopted Gemini's second-wave L892 finding. The previous `"Or omit initial so the entity starts unset."` was accurate for create but misleading on the update path where the merge at L3079-3084 preserves the existing value when `initial` is None. Reworded to `"Or omit initial to use the default or existing value."` so the user-facing error covers both branches the validator is reachable from.
- **Third-wave nit on the same suggestion declined.** The L897 follow-up proposed a three-entry suggestions array including a Home Assistant documentation link and an explicit "new options may leave existing initial out" branch. Declined: sibling validators in the module (`_validate_input_select_options`, `_validate_datetime_has_date_or_time` after its first-wave fix) ship one suggestion without a doc-link, and the edge case is reachable from suggestion #1 ("Pick an `initial` value that's in `options`"). Held to the sibling pattern.

**Problems Encountered:**
- None during implementation. Local test loops stayed green across both review iterations.

**Suggested Improvements (post-merge):**
- **Sibling-validator audit on other update paths.** The same create-only-guard pattern that motivated this PR may exist elsewhere in `tools_config_helpers.py` (e.g. `counter`, `timer`, `input_text` ranges). The numeric-range guard already runs on update via `_validate_numeric_range`; the remaining per-type fields would benefit from the same audit. Out-of-scope here to keep the PR narrowly aligned with issue #1292.

**Review by kingpanther13 — CHANGES_REQUESTED** [review]
Clean refactor — extract-and-share is faithful, merge ordering puts resolution before validation, ToolError propagation is intact through the outer guard at L3387. A few things to address before merge.

## Requested changes

**1. Strip issue-number refs from docstrings/comments.** Per AGENTS.md "comments shouldn't reference the current task/fix/issue — belongs in the PR description." The `#1150` / `#1292` refs in `_validate_initial_in_options` (L867-885) and `_validate_datetime_has_date_or_time` (L887-898), in the call-site comments at L2459-2461 / L3087-3091 / L3187-3191, and in the test class docstrings at L181-187 / L305-311 will read as rot once the migration framing fades. The durable WHY is the parity invariant — keep that, drop the issue numbers.

**2. Pick one shape strategy for `options`.** `_validate_initial_in_options` declares `options: list[Any]` but the body still does `isinstance(options, list)`. Sibling `_validate_input_select_options` at L869-908 uses `options: Any` + an internal shape check — match it, or trust the type and drop the runtime guard. The current "typed-narrowly but defended anyway" reads as confused intent.

**3. The `isinstance(options, list)` guard is unreachable.** Both call sites feed lists (caller path goes through `parse_string_list_param`; existing path defaults to `[]`). If you keep the guard as defense-in-depth, add a direct unit test pinning the contract. Otherwise drop it — unreachable guards drift into "why is this here?" over time.

## While you're in there

- **Missing test edges.** `initial=""` against `options=["A","B"]` (empty string treated as set, should reject), `options=[]` on update with any `initial`, and a happy-path update where caller passes `has_date=None`/`has_time=None` against an existing `(False, True)` config. Cheap to add and locks the contract.
- **Hardcoded `"input_select"` in error context** (`_validate_initial_in_options` L887). Fine today — both callers are input_select branches — but a future reuse on another helper type would emit misleading context. A `helper_type` param costs one line.
- **Collapse `_wire_existing_config` into `_wire_default_ws`.** It's a thin wrapper that differs only by passing `existing_config=`. Adding a default `existing_config=None` param to the original is one helper instead of two.
- **PR description should note the create-branch suggestion text change.** "Or omit `initial` so the entity starts unset" → "Or omit `initial` to use the default or existing value." Intentional generalization for shared use, but it's a user-visible message shift on the create path — worth surfacing so it doesn't get lost.

**Review by Patch76 — COMMENTED** [review]
Addressed in 944964a — mapping each item:

## Requested

**1. Strip issue-number refs from docstrings/comments/tests.** Done. Removed every `#1150` / `#1292` reference the PR introduced: validator docstrings (`_validate_initial_in_options`, `_validate_datetime_has_date_or_time`), the four call-site comments in the create/update branches, the section banner above the new test classes, both test class docstrings (`TestInputSelectInitialInOptions`, `TestInputDatetimeHasDateOrTime`), and the inline `--- Create-side / Update-side ---` markers. The durable WHY — "parity invariant across create and update" — stays in prose. Pre-existing `#1150` markers on Bug 9 / Bug 13 / Bug 17 sections outside this PR's diff are left untouched.

**2. Pick one shape strategy for `options`.** Done. `_validate_initial_in_options` signature changes from `options: list[Any]` to `options: Any`, matching the sibling `_validate_input_select_options` at L929. The `isinstance(options, list)` early-return stays inside, mirroring the same defensive pattern — invariant gate, not type contract.

**3. The `isinstance(options, list)` guard is unreachable from current callers.** Kept as defense-in-depth (matches sibling) and pinned with a direct contract test: `TestValidateInitialInOptionsShapeGuard` exercises the helper with `None` / `"not a list"` / `{"a": 1}` options — all return silently rather than raising `TypeError` on the membership check.

## While you're in there

- **Missing test edges.** Added:
  - `test_empty_string_initial_rejected_against_non_empty_options` — `initial=""` against `options=["A","B"]`: the empty string is a *set* value (not `None`), so the truthy-only inline check the pre-helper code used would have silently dropped it. The new validator rejects.
  - `test_empty_options_with_any_initial_rejected` — `options=[]` with any non-`None` initial rejects.
  - `test_update_rejects_initial_empty_string_against_existing_options` — end-to-end mirror through the update path.
  - `test_update_happy_path_omits_both_against_one_remaining` — caller passes neither `has_date` nor `has_time` against existing `(False, True)`; the merge leaves the existing state untouched and the guard must not false-positive. Locks the no-op-ish update happy path.

- **Hardcoded `"input_select"` in error context.** Parameterized — `_validate_initial_in_options(options, initial, helper_type="input_select")`. Default preserves current call-site behaviour; a future reuse on another helper type now threads the right identifier into both the error message and `_simple_helper_error_context`. Locked by `test_helper_type_param_threads_to_error_context`.

- **Collapse `_wire_existing_config` into `_wire_default_ws`.** Done. `_wire_default_ws` now accepts `existing_config: dict[str, Any] | None = None`; the seven `_wire_existing_config(...)` call sites migrate to the unified helper. One helper instead of two thin wrappers.

- **PR description note on the create-branch suggestion text change.** Added a "User-visible message shift on the create path" paragraph: *"Or omit `initial` so the entity starts unset"* → *"Or omit `initial` to use the default or existing value."* — intentional generalization for the dual-branch use, accurate on both paths.

---

**Local verification (Alpine, Python 3.13 musl):**
- `uv run pytest tests/src/unit/`: **2459 passed, 1 skipped** (numpy by-design)
- `tests/src/unit/test_helper_input_validation.py`: **44/44 passed** (was 35; +9 new tests)
- `uv run ruff check src/ tests/src/unit/`: clean
- `uv run mypy src/ha_mcp/tools/tools_config_helpers.py`: clean

**Review by kingpanther13 — APPROVED** [review]
All requested changes addressed in c66bd52b. Shape strategy resolved, helper_type parameterized, _wire_default_ws collapse clean, test edges covered.
